### PR TITLE
add rust pub-sub example

### PR DESCRIPTION
**Pull request by witterlee**
**Overview**
Why do we need this feature? What are we trying to accomplish?
Rust pubsub example
**Requirements**
What requirements are necessary to consider this problem solved?
More Rust examples should be written
**Considered Solutions**
What solutions were considered?

**Recommended Solution**
What solution are you recommending? Why?

**Considerations**
What impact will this change have on security, performance, users (e.g. breaking changes) or the team (e.g. maintenance costs)?

### CODE REVIEW

**Review by flyq — APPROVED** [review]
Great, need more rust examples.

**Review by DelandPro — APPROVED** [review]
please merge it

File: rust/pub-sub/src/subscriber/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+use ic_cdk::export::{
+    candid::{CandidType, Deserialize},
+    Principal,
+};
+use ic_cdk_macros::*;
+use serde::Serialize;
```
[the remark this answers is not in the document]
Comment (witterlee) [reply rust/pub-sub/src/subscriber/src/lib.rs:6]: accept

File: rust/pub-sub/src/publisher/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+}
+
+#[update]
+async fn publish(counter: Counter) {
+    let subscriber_store = storage::get_mut::<SubscriberStore>();
+    for (k, v) in subscriber_store {
+        if v.topic == counter.topic {
+            let _call_result: Result<(), _> =
+                ic_cdk::api::call::call(*k, "update_count", (&counter,)).await;
```
[the remark this answers is not in the document]
Comment (witterlee) [reply rust/pub-sub/src/publisher/src/lib.rs:38]: cannot use the `?` operator in an async function that returns `()`

File: rust/pub-sub/src/subscriber/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#[derive(Clone, Debug, CandidType, Serialize, Deserialize)]
+struct Subscriber {
+    topic: String,
+}
+
+#[update]
+async fn setup_subscribe(publisher_id: Principal, topic: String) {
+    let subscriber = Subscriber { topic };
+    let _call_result: Result<(), _> =
+        ic_cdk::api::call::call(publisher_id, "subscribe", (subscriber,)).await;
```
[the remark this answers is not in the document]
Comment (witterlee) [reply rust/pub-sub/src/subscriber/src/lib.rs:25]: cannot use the `?` operator in an async function that returns `()`

File: rust/pub-sub/src/subscriber/src/subscriber.did [hunk 4]
```diff
@@ -0,0 +1,12 @@
+type Counter = variant {    
+    topic : text;
+    value:nat8;
```
[the remark this answers is not in the document]
Comment (witterlee) [reply rust/pub-sub/src/subscriber/src/subscriber.did:3]: accept

File: rust/pub-sub/src/subscriber/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+[package]
```
[the remark this answers is not in the document]
Comment (witterlee) [reply rust/pub-sub/src/subscriber/Cargo.toml:1]: accept

File: rust/pub-sub/src/publisher/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    Principal,
+};
+use ic_cdk::storage;
+use ic_cdk_macros::*;
+use serde::Serialize;
+use std::collections::BTreeMap;
+
+type SubscriberStore = BTreeMap<Principal, Subscriber>;
+
+#[derive(Clone, Debug, CandidType, Serialize, Deserialize)]
```
[the remark this answers is not in the document]
Comment (witterlee) [reply rust/pub-sub/src/publisher/src/lib.rs]: accept

File: .gitignore [hunk 7]
```diff
@@ -1,3 +1,19 @@
 .dfx/
 build/
 node_modules/
+
+
+# Generated by Cargo
+# will have compiled files and executables
+debug/
+target/
```
[the remark this answers is not in the document]
Comment (witterlee) [reply .gitignore:9]: accept

File: rust/pub-sub/src/publisher/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenyan-dfinity) [reply rust/pub-sub/src/publisher/src/lib.rs:38]: You can either return `Result` or at least `unwrap`. Not consuming the result value is bad.

**Review by krpeacock — APPROVED** [review]
@chenyan-dfinity I feel like this is worth accepting as-is and we can always shore up any small details later

**Pull request merged**